### PR TITLE
#1020 Test to close server when replicator point to none existed uri

### PR DESCRIPTION
**Pull request by Sandychuang8**
Add a test case to test when server is closed when replicator point to none existed uri

---
This change is [Reviewable](https://reviewable.io/reviews/couchbase/couchbase-lite-net/1030)

### CODE REVIEW

File: src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs [hunk 1]
```diff
@@ -70,9 +70,35 @@ public ReplicatorTest()
             ReopenDB();
             Database.Delete("otherdb", Directory);
             _otherDB = OpenDB("otherdb");
+            Database.SetLogLevel(LogDomain.Replicator, LogLevel.Verbose);
         }
 
-        #if COUCHBASE_ENTERPRISE
+        [Fact]
+        public async Task TestCloseServerWhenReplicatorExistsAsync()
```
Comment (borrrden) [inline src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs]: The point of the test is to check that the replicator stops when it cannot connect to a non-existent endpoint so I'd change the name to reflect that

File: src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs [hunk 2]
```diff
@@ -70,9 +70,35 @@ public ReplicatorTest()
             ReopenDB();
             Database.Delete("otherdb", Directory);
             _otherDB = OpenDB("otherdb");
+            Database.SetLogLevel(LogDomain.Replicator, LogLevel.Verbose);
```
Comment (borrrden) [inline src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs]: This was meant to be temporary.  If left in, all the logs for the rest of the test suite will be at this level

File: src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs [hunk 3]
```diff
@@ -70,9 +70,35 @@ public ReplicatorTest()
...
+            var targetEndpoint = new URLEndpoint(new Uri("ws://192.168.0.11:4984/app"));
+            var config = new ReplicatorConfiguration(Db, targetEndpoint);
+            using (var repl = new Replicator(config))
+            {
+                repl.Start();
+                var count = 0;
+
+                this.Invoking(x => ReopenDB())
+                    .ShouldThrow<CouchbaseLiteException>(
+                        "because the database cannot be closed while replication is running");
```
Comment (borrrden) [inline src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs]: This assertion is not essential to the test, but there is no harm in leaving it in

File: src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs [hunk 4]
```diff
@@ -70,9 +70,35 @@ public ReplicatorTest()
...
+            
+                while (count++ > 20 && repl.Status.Activity != ReplicatorActivityLevel.Stopped)
+                {
+                    WriteLine($"Replication status still {repl.Status.Activity}, waiting for stopped...");
+                    await Task.Delay(500);
+                }
+
+                if(repl.Status.Activity == ReplicatorActivityLevel.Stopped)
+                    WriteLine($"Replication status still {repl.Status.Activity}, stopped");
```
Comment (borrrden) [inline src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs]: This should be an assertion, not just a WriteLine.  The test should fail if the replication does not close.  Note that in the situation we are going to cover, retries will be attempted so it will take quite a while for the replication to stop.  I'll catch up with you about that offline.

File: src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs [hunk 5]
```diff
@@ -70,31 +70,26 @@ public ReplicatorTest()
...
             Database.Delete("otherdb", Directory);
             _otherDB = OpenDB("otherdb");
-            Database.SetLogLevel(LogDomain.Replicator, LogLevel.Verbose);
+            //uncomment the code below when you need to see more detail log
+            //Database.SetLogLevel(LogDomain.Replicator, LogLevel.Verbose);
         }
 
         [Fact]
-        public async Task TestCloseServerWhenReplicatorExistsAsync()
+        public async Task TestToCheckReplicatorStopsWhenAttemptToConnectToInvalidEndpoint()
```
Comment (borrrden) [inline src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs]: This name is a bit long and I don't think it will show up in the test explorer without making it really wide.  How about `TestReplicatorStopsWhenEndpointInvalid`

File: src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs [hunk 6]
```diff
@@ -70,31 +70,26 @@ public ReplicatorTest()
...
-            
                 while (count++ > 20 && repl.Status.Activity != ReplicatorActivityLevel.Stopped)
                 {
                     WriteLine($"Replication status still {repl.Status.Activity}, waiting for stopped...");
                     await Task.Delay(500);
                 }
 
-                if(repl.Status.Activity == ReplicatorActivityLevel.Stopped)
-                    WriteLine($"Replication status still {repl.Status.Activity}, stopped");
+                count.Should().BeLessThan(20);
```
Comment (borrrden) [inline src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs]: Right now the answer is clear as to why the count should be less than 20, but in the future we will probably forget, so it is best to add the optional reason into this assert:

`count.Should().BeLessThan(20, "because otherwise the replicator never stopped");`

This creates a single left to right readable line that explains what the assert is for.
